include/vcpkg/base/parallel-algorithms.h

```diff
@@ -0,0 +1,66 @@
+#pragma once
+
+// At this time only MSVC supports parallel algorithms by default
+// Gcc needs tbb as a runtime dependency
+// Clang doesn't support parallel algorithms at all
+
+#if defined(_MSC_VER)
+#include <algorithm>
+#include <execution>
+#include <mutex>
+
+#define vcpkg_parallel_for_each(BEGIN, END, CB) std::for_each(std::execution::par, BEGIN, END, CB)
+#define vcpkg_par_unseq_for_each(BEGIN, END, CB) std::for_each(std::execution::par_unseq, BEGIN, END, CB)
+
+#else
+#include <atomic>
+#include <future>
+#include <vector>
+
+namespace vcpkg
+{
+    template<class It, class F>
+    void parallel_for_each(It begin, It end, F&& cb)
+    {
+        if (begin == end)
+        {
+            return;
+        }
+        if (begin + 1 == end)
+        {
+            return cb(std::move(*begin));
+        }
+
+        const auto thread_count = std::thread::hardware_concurrency() * 2;
+        const ptrdiff_t work_count = std::distance(begin, end);
+        const auto num_threads = static_cast<size_t>(
+            std::max(static_cast<ptrdiff_t>(1), std::min(static_cast<ptrdiff_t>(thread_count), work_count)));
+
+        std::vector<std::future<void>> workers;
+        workers.reserve(num_threads);
+
+        std::atomic_size_t next{0};
+        auto work = [&]() {
+            size_t i;
+            while (i = next.fetch_add(1, std::memory_order_relaxed), i < static_cast<size_t>(work_count))
+            {
+                cb(std::move(*(begin + i)));
+            }
+        };
+
+        for (size_t i = 0; i < num_threads - 1; ++i)
+        {
+            workers.emplace_back(std::async(std::launch::async, work));
```

Comment:
`launch::async` with `future<void>` is basically the same as using `std::thread`, I think we should do that?

Reply:
I think the benefit of `std::async` is that it maybe uses a thread pool

Reply:
> I think the benefit of `std::async` is that it maybe uses a thread pool

It can't use a thread pool with `launch::async`.

Reply:
https://github.com/microsoft/STL/blob/6735beb0c2260e325c3a4c971ec5b75427e9305f/stl/inc/future#L662
and 
> Use tasks (the [concurrency::task](https://learn.microsoft.com/en-us/cpp/parallel/concrt/reference/task-class?view=msvc-170) class) when you write asynchronous code. The task class uses the Windows ThreadPool as its scheduler, not the Concurrency Runtime.

From https://learn.microsoft.com/en-us/cpp/parallel/concrt/task-parallelism-concurrency-runtime?view=msvc-170

Reply:
> https://github.com/microsoft/STL/blob/6735beb0c2260e325c3a4c971ec5b75427e9305f/stl/inc/future#L662

This is nonconforming: http://eel.is/c++draft/futures.async#3.1

and if I ever get to actually ship the rewritten `<future>` I did in 2016 this nonconformance got fixed.